core/src/al/miniaudio_device.cpp

```diff
@@ -0,0 +1,238 @@
+}
+
+void MiniaudioSource::setVelocity(const glm::vec3& velocity)
+{
+    ma_sound_set_velocity(&mSound, velocity.x, velocity.y, velocity.z);
+}
+
+void MiniaudioSource::setGain(float gain)
+{
+    ma_sound_set_volume(&mSound, gain);
+}
+
+void MiniaudioSource::setPitch(float pitch)
+{
+    ma_sound_set_pitch(&mSound, pitch);
+}
+
+void MiniaudioSource::setLooping(bool looping)
+{
+    ma_sound_set_looping(&mSound, static_cast<ma_bool32>(looping));
+}
+
+void MiniaudioSource::setRelative(bool relative)
+{
+    relative ? ma_sound_set_positioning(&mSound, ma_positioning_relative)
+             : ma_sound_set_positioning(&mSound, ma_positioning_absolute);
+}
+
+void MiniaudioSource::setMaxDistance(float maxDistance)
+{
+    ma_sound_set_max_distance(&mSound, maxDistance);
+}
+
+void MiniaudioSource::setMinDistance(float minDistance)
+{
+    ma_sound_set_min_distance(&mSound, minDistance);
+}
+
+void MiniaudioSource::setCone(float innerAngle, float outerAngle, float outerGain = 1.0F)
+{
+    ma_sound_set_cone(&mSound, innerAngle, outerAngle, outerGain);
+}
+
+void MiniaudioSource::setConeDirection(const glm::vec3& direction)
+{
+    ma_sound_set_direction(&mSound, direction.x, direction.y, direction.z);
+}
+
+void MiniaudioSource::play()
+{
+    if (ma_sound_start(&mSound) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to start sound.");
+        abort();
+    }
+}
+
+MiniaudioDevice::MiniaudioDevice()
+{
+    // Initialize miniaudio context.
+    if (ma_context_init(nullptr, 0, nullptr, &mContext) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize miniaudio context.");
+        abort();
+    }
+
+    // Initialize miniaudio engine
+    if (ma_engine_init(nullptr, &mEngine) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize miniaudio engine.");
+        abort();
+    }
+
+    // Configure the device.
+    ma_device_config deviceConfig = ma_device_config_init(ma_device_type_playback);
+    deviceConfig.playback.format = ma_format_f32; // Set to ma_format_unknown to use the device's native format.
+    deviceConfig.playback.channels = 2;           // Set to 0 to use the device's native channel count.
+    deviceConfig.sampleRate = 48000;              // Set to 0 to use the device's native sample rate.
+
+    // Initialize the audio device.
+    if (ma_device_init(&mContext, &deviceConfig, &mDevice) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize audio device.");
+        ma_context_uninit(&mContext);
+        abort();
+    }
+
+    ma_device_start(&mDevice);
+}
+
+MiniaudioDevice::~MiniaudioDevice()
+{
+
+    ma_device_uninit(&mDevice);
+    ma_context_uninit(&mContext);
+}
+
+void MiniaudioDevice::enumerateDevices(std::vector<std::string>& devices)
+{
+    ma_context context;
+    if (ma_context_init(nullptr, 0, nullptr, &context) != MA_SUCCESS)
+    {
+        CUBOS_CRITICAL("Failed to initialize audio context.");
+        abort();
+    }
```

Comment:
It feels like we should do this a single time per application. In the docs it says somewhere that some platforms only support one context to be created at the same time, which means:
```cpp
auto audioDevice = AudioDevice::create(); // or smth, i don't remember the exact syntax; leads to context being inited
AudioDevice::enumerateDevices(devices); // crash, another context has been inited
```

Maybe this AudioDevice abstraction class should be split into an `AudioContext` and an `AudioDevice`?
You create an `AudioContext`, and through it, you can get the list of devices, and get an handle for each of them (`AudioDevice`).